File: bench/src/otplike/bench/runner.clj
(ns otplike.bench.runner
  "Main entry point for otplike benchmarks."
  (:require [otplike.bench.common :as c]
            [otplike.bench.spawn :as spawn]
            [otplike.bench.memory :as memory]
            [otplike.bench.messaging :as messaging]
            [otplike.bench.leaks :as leaks])
  (:gen-class))

(defn -main [& args]
  (c/print-system-info "otplike")
  (println)
  
  (let [mode (first args)]
    (case mode
      "spawn"     (spawn/run-all)
      "memory"    (memory/run-all)
      "messaging" (messaging/run-all)
      "scaling"   (memory/run-scaling)
      "leaks"     (leaks/run-all)
      "leaks-quick" (leaks/run-quick)
      "leaks-nested" (leaks/run-nested-async-stress)
      ;; Default: run all
      (do
        (spawn/run-all)
        (println)
        (memory/run-all)
        (println)
        (messaging/run-all))))
  
  (println)
  (println "========================================")
  (println "otplike benchmarks complete")
  (println "========================================")
  (flush)
  (System/exit 0))

File: bench/src/elixir/lib/bench_elixir/runner.ex
defmodule BenchElixir.Runner do
  @moduledoc """
  Main entry point for Elixir benchmarks.
  """

  alias BenchElixir.Bench.{Common, Spawn, Messaging, Memory}

  def main(args \\ []) do
    # Start os_mon for memory info
    Application.ensure_all_started(:os_mon)

    Common.print_system_info("Elixir/BEAM")
    IO.puts("")

    case args do
      ["spawn"] ->
        Spawn.run_all()

      ["memory"] ->
        Memory.run_all()

      ["messaging"] ->
        Messaging.run_all()

      ["scaling"] ->
        Memory.run_scaling()

      [] ->
        # Default: run all
        Spawn.run_all()
        IO.puts("")
        Memory.run_all()
        IO.puts("")
        Messaging.run_all()

      [other] ->
        IO.puts("Unknown mode: #{other}")
        IO.puts("Usage: mix run -e 'BenchElixir.Runner.main([\"mode\"])'")
        IO.puts("Modes: spawn, memory, messaging, scaling")
        IO.puts("Or run without args for all benchmarks")
    end

    IO.puts("")
    IO.puts("========================================")
    IO.puts("Elixir/BEAM benchmarks complete")
    IO.puts("========================================")
  end
end
